### ISSUE TITLE
APKAM - keys verb implementation and server impl for spec changes

### ISSUE THREAD

**Issue by murali-shris** [atsign-foundation/at_server#1437]
### Is your feature request related to a problem? Please describe.

Implement keys verb in server as per APKAM spec
### Initial bootstrap - onboard FirstApp
- FirstApp does a CRAM authentication
- FirstApp enrolls itself and its APKAM
  - Cut its APKAM encryption keypair (or get the APKAM public key from a TPM / Secure Element)
  - Enrolls:
    ```
    enroll:request:app:<appName>:device:<deviceName>
    :namespaces:<one,r;two,rw;three,r[;...]
    :apkamPublicKey:<apkamPublicKey>
    ```
  - automatically approved because app has authenticated with CRAM secret
  - and given rw access to the `__manage` namespace
- FirstApp disconnects, creates a new connection and does APKAM authentication
  `from:@atSign`, and then
  `pkam:app:<appName>:device:<deviceName>:<signedChallenge>`
- Create default encryption keypair
  - FirstApp cuts a default encryption keypair and gives it a name e.g. firstKey
  - FirstApp stores the default encryption keypair's public key
    ```
    keys:put:public
      :keyName:firstKey:namespace:__global
      :keyType:<RSA-2048|eccxyz|etc>
      :<public key base64 encoded>
    ```
    - public key is stored in `public:<keyName>.__public_keys.__global@atSign`
      - with `keyType` in the metadata
    - For backwards compatibility, also stores `public:publickey@atSign` as a
      reference to `public:<keyName>.__public_keys.__global@atSign`
  - FirstApp encrypts the default encryption keypair's private key using FirstApp's
  apkamPublicKey, and stores it for later use by FirstApp
    ```
    keys:put:private:app:<appName>:device:<deviceName>
      :keyName:firstKey:namespace:__global
      :<encryptedEncryptionPrivateKey>
    ```
    - appName and deviceName must match what was used in PKAM
    - Creates key `private:<appName>.<deviceName>.<keyName>.__private_keys.__global@atSign`
      - with value being `<encryptedEncryptionPrivateKey>`
- Create a default 'self' encryption key
  - FirstApp cuts new symmetricSelfEncryptionKey (e.g. an AES key)
  - Encrypts it using default global encryption public key e.g. firstKey
  - Stores it
    ```
    keys:put:self
    :keyName:<keyName>:namespace:__global
      :keyType:<AES-256|XYZ-128|etc>:encryptionKeyName:<keyName e.g. firstKey>
      :<encryptedSymmetricSelfEncryptionKey>
    ```
    - stored in `<keyName>.__self_keys.__global@atSign`
    - with `keyType` and `encryptionKeyName` in the metadata

### Subsequent runs of FirstApp
- Do APKAM authentication
- Retrieve all encryption keypairs' private keys
  - `keys:get:private`
    - Retrieves everything from the `__private_keys.__global` namespace
    - Retrieves everything from `__private_keys.$namespace` for each $namespace to
      which this app has access
  - (Recall that these are encrypted with FirstApp's APKAM public key)
- Retrieve all self encryption keys
  - `keys:get:self`
    - Retrieves everything from the `__self_keys.__global` namespace
    - Retrieves everything from `__self_keys.$namespace` for each $namespace to
      which this app has access
  - (Recall that these are encrypted with some encryption keypair's public key)

### Describe the solution you'd like

* add syntax for keys verb in at_commons
* implement verb builder in at_commons
* implement verb handler in at_server

**Comment by murali-shris** [comment]
@gkc Adding a metadata to persistence metadata object requires changes in lot of files. Is it fine if store the key type in value as a part of json instead of metadata?
e.g.
key: public:myencryptionPublicKey.__public_keys.__global@alice 

value:{value: <base64encoded_pubkey>, keyType: rsa2048}

**Comment by gkc** [comment]
> @gkc Adding a metadata to persistence metadata object requires changes in lot of files. Is it fine if store the key type in value as a part of json instead of metadata? e.g. key: public:myencryptionPublicKey.__public_keys.__global@alice
> 
> value:{value: <base64encoded_pubkey>, keyType: rsa2048}

Yes that's fine for sure

**Comment by murali-shris** [comment]
@gkc 
private keys with __global namespace
I can lookup the key from normal pkam authenticated connection.
@ alice @ llookup:private:wavi.pixel.secretKey.__private_keys.__global@alice
data:{"value":"abcd1234","keyType":"aes"}


With APKAM authenticated connection I am getting this since the APKAM has access only to .buzz namespace
@ alice @ llookup:private:wavi.pixel.secretKey.__private_keys.__global@alice
error:AT0009-UnAuthorized client in request : Enrollment Id: a514d324-45a2-4f36-80e9-08b83a454c0e is not authorized for local lookup operation
If I allow this APKAM connection to access __global private keys...other APKAM connections should not see this key right?

**Comment by murali-shris** [comment]
PR under review
https://github.com/atsign-foundation/at_server/pull/1463

**Comment by murali-shris** [comment]
Remove reference from the current implementation and store the default encryption public key value for both keys
```
public:encryption_<enrollmentId>.__public_keys.__global@atSign
public:publickey@atSign
```

**Comment by murali-shris** [comment]
Change enroll syntax and enrol verb handler impl as per recent spec changes
https://github.com/atsign-foundation/at_protocol/blob/trunk/decisions/2023-01-pkam-per-app-and-device.md

**Comment by murali-shris** [comment]
Implementation changes:
* Modify enroll verb to accept a json for params.  Added new class in at_commons EnrollParams.dart
* Remove reference key for publickey from previous impl
* In enroll verb handler in order to auto approve enroll request, connection should be cram authenticated. (was cram or pkam authentication in previous impl)
* Rename enrollApprovalId to enrollmentId for naming consistency
* Add encrypted default private encryption key and encrypted default self encryption key to keystore when enroll request is approved. both the keys are encrypted with enrollment's apkam symmetric key
* Store apkamPublicKey in  public:appName.deviceName.pkam.__pkams.__public_keys@ <atsign> when enroll request is approved
* Removed EnrollNamespace object from previous impl and replace it with Map<String,String> (key - namespace, value -  access) for ease of retrieval
* For approval notification, store encrypted apkam symmetric key in notification value.Approving app will decrypt using default encryption private key 

Changes done so far in 
https://github.com/atsign-foundation/at_server/tree/apkam_spec_changes
https://github.com/atsign-foundation/at_libraries/tree/apkam_enc_dec_changes
Sample client to test approval flow
https://gist.github.com/murali-shris/55d1d67a4510bed1363d811140592444

**Comment by murali-shris** [comment]
@gkc please review these steps and whether the output is as expected per the spec

End to end test - manual steps
**First client**
1) Start a secondary with cram secret
2) Cram auth from openssl
3)  Use a dart client to encrypt default encryption private key and default self encryption key
var aliceDefaultEncKey = at_demos.encryptionPrivateKeyMap['@ alice'];
      var aliceSelfEncKey = at_demos.aesKeyMap['@ alice'];
      var aliceApkamSymmetricKey = at_demos.apkamSymmetricKeyMap['@ alice'];
      var encryptedDefaultEncPrivateKey = EncryptionUtil.encryptValue(aliceDefaultEncKey!, aliceApkamSymmetricKey!);
      print(encryptedDefaultEncPrivateKey);
      var encryptedSelfEncKey = EncryptionUtil.encryptValue(aliceSelfEncKey!, aliceApkamSymmetricKey!);
      print(encryptedSelfEncKey);
4) Send the below enroll request      
```
enroll:request:{"appName":"wavi","deviceName":"pixel","namespaces":{"wavi":"rw"},"encryptedDefaultEncryptedPrivateKey":"<encryptedDefaultEncPrivateKey from step 3>","encryptedDefaultSelfEncryptionKey":"<encryptedSelfEncKey from step 3>","apkamPublicKey":"<use pkam public key of alice from at_demo_data>"}
```
Enrollment gets auto approved since connection is cram authenticated
```
data:{"enrollmentId":"e8e69f4a-2336-4f90-8235-f0a66b5713db","status":"success"}
```
5) Disconnect openssl client
6) Connect to secondary using openssl client and pkam auth using the enrollmentId from step 4
```
@pkam:enrollmentId:e8e69f4a-2336-4f90-8235-f0a66b5713db:<signature>
data:success
@alice@
```
 6.1) scan doesn't return keys from __manage namespace.   
```
@alice@scan
data:["public:signing_publickey@alice"]
```
 6.2) enroll:list return the current enrollment. Since the client was cram authenticated during enrollment, it has access to __manage namespace
```
data:{"e8e69f4a-2336-4f90-8235-f0a66b5713db.new.enrollments.__manage@alice":{"appName":"wavi","deviceName":"pixel","namespace":{"wavi":"rw","__manage":"rw"}}}
```
 6.3) llookup on above key fails
```
@alice@llookup:e8e69f4a-2336-4f90-8235-f0a66b5713db.new.enrollments.__manage@alice
error:AT0009-UnAuthorized client in request : Enrollment Id: e8e69f4a-2336-4f90-8235-f0a66b5713db is not authorized for local lookup operation on the key: e8e69f4a-2336-4f90-8235-f0a66b5713db.new.enrollments.__manage@alice
```
 6.5) Get the private and self keys
```
@alice@keys:get:self
data:["e8e69f4a-2336-4f90-8235-f0a66b5713db.default_self_enc_key.__manage@alice"]
@alice@keys:get:private
data:["e8e69f4a-2336-4f90-8235-f0a66b5713db.default_enc_private_key.__manage@alice"]
```
 6.6) __manage keys can be looked up using keys:get:keyName:
```
@alice@keys:get:keyName:e8e69f4a-2336-4f90-8235-f0a66b5713db.default_self_enc_key.__manage@alice
data:{"value":"EdHYR+Rol3v1pAigsuSSkOTAkcse3wHXIX9b4tZd7SBrHLXJnZt2EqLur5pIKKjD"}
@alice@keys:get:keyName:e8e69f4a-2336-4f90-8235-f0a66b5713db.default_enc_private_key.__manage@alice
data:{"value":"Ksq6db1VpgDFiT2E3......."}
```
7) pkam without enrollmentId should work for first onboarded client for backward compatibility
```
@from:@alice
data:_54c2b527-c022-4a06-8dc3-0d9a81942a98@alice:95f59c10-3ff0-485e-baf3-447449f369bf
@pkam:Zusgp39QzRdOYVqZLzO505s1ct6JtCjrFtN0oYeXaNm5f4qNTBfU01rHxZcPkOBRxHuYtsTHe+/7aSYc3IV1K/jzWPC7d/XbiLAlWzv/CY3TxomQuA2oVy7Ro/ADvm8SJcN0UFbtG/mYbtTVPdSST2m1ep2ZNJoex+lK3i76lX1ZhIMJkp2GvttlcF73HzUtEvL+Qj9tC0SgZMM8czCNmn8ZsY8zYNvJ8Pb+B1M/H1ltbe7txPDHXP2cnUhtFD02pV06uqTSINjjGBXWyXz0cwb/2KQ7pKHdbWT8LrNiabSRU5LZwmL13gUzS1FFOGOTU87mGy+e3oFtVk1D9wdSVg==
data:success
@alice@
```
**Second client**
1) Open a terminal and run sample client which can approve or deny new enrollment requests
https://gist.github.com/murali-shris/55d1d67a4510bed1363d811140592444
2) Get totp from enrolled client
```
@alice@totp:get      
data:949582
```
3) Connect to secondary through openssl and send enroll request for new client
```
murali@murali-Latitude-3460:~/softwares/pycharm-community-2017.3.1/bin$ openssl s_client -quiet -connect localhost:6400
depth=2 C = US, O = Internet Security Research Group, CN = ISRG Root X1
verify return:1
depth=1 C = US, O = Let's Encrypt, CN = R3
verify return:1
depth=0 CN = vip.ve.atsign.zone
verify return:1
@enroll:request:{"appName":"buzz","deviceName":"pixel","namespaces":{"buzz":"rw"},"totp":949582, "encryptedAPKAMSymmetricKey":"RbfCJO...."}
data:{"enrollmentId":"1a901b8e-0e89-4b09-9e94-d7b06a12d818","status":"pending"}
```
4) Sample client in step 1) will get a notification for the enrollment
```
Approve enrollmentId 1a901b8e-0e89-4b09-9e94-d7b06a12d818?
```

4.1) Approve the enrollment

```
Approve enrollmentId 1a901b8e-0e89-4b09-9e94-d7b06a12d818?
yes
approved?: yes
decrypted apkam symmetric key: 8QHaELcv0QBSBLTc/dpNHenbuZeT95+0GDwq3Lugbt8=
enroll request to server: enroll:approve:{"enrollmentId":"1a901b8e-0e89-4b09-9e94-d7b06a12d818","encryptedDefaultEncryptedPrivateKey":"DF2EjCyI.....,"encryptedDefaultSelfEncryptionKey":"N0bmvn...."}

enrollResponse: data:{"status":"approved","enrollmentId":"1a901b8e-0e89-4b09-9e94-d7b06a12d818"}
```
5) Try pkam auth with enrollmentId
```
@from:@alice
data:_89cf667c-9d35-406b-85af-6d344ca581d3@alice:7683a23c-36ba-4504-a2f4-5e11467377ad
@pkam:enrollmentId:1a901b8e-0e89-4b09-9e94-d7b06a12d818:G5oieA/dNIpuu0c1QAhJTxupJwxxEV9I+c1PTIR+/ZGE6zmTnUz/iSqC6w8dRq4e1yNoaJCla8C6BcJGTbwqJdTBomZZnJMicVbLtmyi8Iixa1gQ1V/nqwHjK8SduYQhh4JPjHapnqL+xzfbFAfMTY5eDmnvYyvuUmy6BqS6Nwl503SN7sqThbVpvyIjI5cK/XOx4pJHbWD7uaKcTo9qIiCntErCdcMRKoPcxwVhz1olnFzh788Fk5yEzRv/492Olm27gOS8PCSNhtOeCp0aB/FxbyLnMjk4qSs0D2rQD1ckf0/6YNmGmKHvLu2kECNaSayoQIt7Rvp+I9J5WXNaNw==
data:success
@alice@
```
6) Get the private and self keys
```
@alice@keys:get:self
data:["1a901b8e-0e89-4b09-9e94-d7b06a12d818.default_self_enc_key.__manage@alice"]
@alice@keys:get:private
data:["1a901b8e-0e89-4b09-9e94-d7b06a12d818.default_enc_private_key.__manage@alice"]
```
**First client**
1) First client has rw access to __manage namespace. it can see all enrollments and __manage keys
```
@alice@enroll:list
data:{"1a901b8e-0e89-4b09-9e94-d7b06a12d818.new.enrollments.__manage@alice":{"appName":"buzz","deviceName":"pixel","namespace":{"buzz":"rw"}},"e8e69f4a-2336-4f90-8235-f0a66b5713db.new.enrollments.__manage@alice":{"appName":"wavi","deviceName":"pixel","namespace":{"wavi":"rw","__manage":"rw"}}}
@alice@keys:get:self
data:["1a901b8e-0e89-4b09-9e94-d7b06a12d818.default_self_enc_key.__manage@alice","e8e69f4a-2336-4f90-8235-f0a66b5713db.default_self_enc_key.__manage@alice"]
@alice@keys:get:private
data:["1a901b8e-0e89-4b09-9e94-d7b06a12d818.default_enc_private_key.__manage@alice","e8e69f4a-2336-4f90-8235-f0a66b5713db.default_enc_private_key.__manage@alice"]
```

**Comment by murali-shris** [comment]
> End to end test - manual steps **First client**
> 
> 1. Start a secondary with cram secret
> 2. Cram auth from openssl
> 3. Use a dart client to encrypt default encryption private key and default self encryption key
>    var aliceDefaultEncKey = at_demos.encryptionPrivateKeyMap['@ alice'];
>    var aliceSelfEncKey = at_demos.aesKeyMap['@ alice'];
>    var aliceApkamSymmetricKey = at_demos.apkamSymmetricKeyMap['@ alice'];
>    var encryptedDefaultEncPrivateKey = EncryptionUtil.encryptValue(aliceDefaultEncKey!, aliceApkamSymmetricKey!);
>    print(encryptedDefaultEncPrivateKey);
>    var encryptedSelfEncKey = EncryptionUtil.encryptValue(aliceSelfEncKey!, aliceApkamSymmetricKey!);
>    print(encryptedSelfEncKey);
> 4. Send the below enroll request
> 
> ```
> enroll:request:{"appName":"wavi","deviceName":"pixel","namespaces":{"wavi":"rw"},"encryptedDefaultEncryptedPrivateKey":"<encryptedDefaultEncPrivateKey from step 3>","encryptedDefaultSelfEncryptionKey":"<encryptedSelfEncKey from step 3>","apkamPublicKey":"<use pkam public key of alice from at_demo_data>"}
> ```
> 
> Enrollment gets auto approved since connection is cram authenticated
> 
> ```
> data:{"enrollmentId":"e8e69f4a-2336-4f90-8235-f0a66b5713db","status":"success"}
> ```
> 
> 5. Disconnect openssl client
> 6. Connect to secondary using openssl client and pkam auth using the enrollmentId from step 4
> 
> ```
> @pkam:enrollmentId:e8e69f4a-2336-4f90-8235-f0a66b5713db:<signature>
> data:success
> @alice@
> ```
> 
> 6.1) scan doesn't return keys from __manage namespace.
> 
> ```
> @alice@scan
> data:["public:signing_publickey@alice"]
> ```
> 
> 6.2) enroll:list return the current enrollment. Since the client was cram authenticated during enrollment, it has access to __manage namespace
> 
> ```
> data:{"e8e69f4a-2336-4f90-8235-f0a66b5713db.new.enrollments.__manage@alice":{"appName":"wavi","deviceName":"pixel","namespace":{"wavi":"rw","__manage":"rw"}}}
> ```
> 
> 6.3) llookup on above key fails
> 
> ```
> @alice@llookup:e8e69f4a-2336-4f90-8235-f0a66b5713db.new.enrollments.__manage@alice
> error:AT0009-UnAuthorized client in request : Enrollment Id: e8e69f4a-2336-4f90-8235-f0a66b5713db is not authorized for local lookup operation on the key: e8e69f4a-2336-4f90-8235-f0a66b5713db.new.enrollments.__manage@alice
> ```
> 
> 6.5) Get the private and self keys
> 
> ```
> @alice@keys:get:self
> data:["e8e69f4a-2336-4f90-8235-f0a66b5713db.default_self_enc_key.__manage@alice"]
> @alice@keys:get:private
> data:["e8e69f4a-2336-4f90-8235-f0a66b5713db.default_enc_private_key.__manage@alice"]
> ```
> 
> 6.6) __manage keys can be looked up using keys:get:keyName:
> 
> ```
> @alice@keys:get:keyName:e8e69f4a-2336-4f90-8235-f0a66b5713db.default_self_enc_key.__manage@alice
> data:{"value":"EdHYR+Rol3v1pAigsuSSkOTAkcse3wHXIX9b4tZd7SBrHLXJnZt2EqLur5pIKKjD"}
> @alice@keys:get:keyName:e8e69f4a-2336-4f90-8235-f0a66b5713db.default_enc_private_key.__manage@alice
> data:{"value":"Ksq6db1VpgDFiT2E3......."}
> ```
> 
> 7. pkam without enrollmentId should work for first onboarded client for backward compatibility
> 
> ```
> @from:@alice
> data:_54c2b527-c022-4a06-8dc3-0d9a81942a98@alice:95f59c10-3ff0-485e-baf3-447449f369bf
> @pkam:Zusgp39QzRdOYVqZLzO505s1ct6JtCjrFtN0oYeXaNm5f4qNTBfU01rHxZcPkOBRxHuYtsTHe+/7aSYc3IV1K/jzWPC7d/XbiLAlWzv/CY3TxomQuA2oVy7Ro/ADvm8SJcN0UFbtG/mYbtTVPdSST2m1ep2ZNJoex+lK3i76lX1ZhIMJkp2GvttlcF73HzUtEvL+Qj9tC0SgZMM8czCNmn8ZsY8zYNvJ8Pb+B1M/H1ltbe7txPDHXP2cnUhtFD02pV06uqTSINjjGBXWyXz0cwb/2KQ7pKHdbWT8LrNiabSRU5LZwmL13gUzS1FFOGOTU87mGy+e3oFtVk1D9wdSVg==
> data:success
> @alice@
> ```
> 
> **Second client**
> 
> 1. Open a terminal and run sample client which can approve or deny new enrollment requests
>    https://gist.github.com/murali-shris/55d1d67a4510bed1363d811140592444
> 2. Get totp from enrolled client
> 
> ```
> @alice@totp:get      
> data:949582
> ```
> 
> 3. Connect to secondary through openssl and send enroll request for new client
> 
> ```
> murali@murali-Latitude-3460:~/softwares/pycharm-community-2017.3.1/bin$ openssl s_client -quiet -connect localhost:6400
> depth=2 C = US, O = Internet Security Research Group, CN = ISRG Root X1
> verify return:1
> depth=1 C = US, O = Let's Encrypt, CN = R3
> verify return:1
> depth=0 CN = vip.ve.atsign.zone
> verify return:1
> @enroll:request:{"appName":"buzz","deviceName":"pixel","namespaces":{"buzz":"rw"},"totp":949582, "encryptedAPKAMSymmetricKey":"RbfCJO...."}
> data:{"enrollmentId":"1a901b8e-0e89-4b09-9e94-d7b06a12d818","status":"pending"}
> ```
> 
> 4. Sample client in step 1) will get a notification for the enrollment
> 
> ```
> Approve enrollmentId 1a901b8e-0e89-4b09-9e94-d7b06a12d818?
> ```
> 
> 4.1) Approve the enrollment
> 
> ```
> Approve enrollmentId 1a901b8e-0e89-4b09-9e94-d7b06a12d818?
> yes
> approved?: yes
> decrypted apkam symmetric key: 8QHaELcv0QBSBLTc/dpNHenbuZeT95+0GDwq3Lugbt8=
> enroll request to server: enroll:approve:{"enrollmentId":"1a901b8e-0e89-4b09-9e94-d7b06a12d818","encryptedDefaultEncryptedPrivateKey":"DF2EjCyI.....,"encryptedDefaultSelfEncryptionKey":"N0bmvn...."}
> 
> enrollResponse: data:{"status":"approved","enrollmentId":"1a901b8e-0e89-4b09-9e94-d7b06a12d818"}
> ```
> 
> 5. Try pkam auth with enrollmentId
> 
> ```
> @from:@alice
> data:_89cf667c-9d35-406b-85af-6d344ca581d3@alice:7683a23c-36ba-4504-a2f4-5e11467377ad
> @pkam:enrollmentId:1a901b8e-0e89-4b09-9e94-d7b06a12d818:G5oieA/dNIpuu0c1QAhJTxupJwxxEV9I+c1PTIR+/ZGE6zmTnUz/iSqC6w8dRq4e1yNoaJCla8C6BcJGTbwqJdTBomZZnJMicVbLtmyi8Iixa1gQ1V/nqwHjK8SduYQhh4JPjHapnqL+xzfbFAfMTY5eDmnvYyvuUmy6BqS6Nwl503SN7sqThbVpvyIjI5cK/XOx4pJHbWD7uaKcTo9qIiCntErCdcMRKoPcxwVhz1olnFzh788Fk5yEzRv/492Olm27gOS8PCSNhtOeCp0aB/FxbyLnMjk4qSs0D2rQD1ckf0/6YNmGmKHvLu2kECNaSayoQIt7Rvp+I9J5WXNaNw==
> data:success
> @alice@
> ```
> 
> 6. Get the private and self keys
> 
> ```
> @alice@keys:get:self
> data:["1a901b8e-0e89-4b09-9e94-d7b06a12d818.default_self_enc_key.__manage@alice"]
> @alice@keys:get:private
> data:["1a901b8e-0e89-4b09-9e94-d7b06a12d818.default_enc_private_key.__manage@alice"]
> ```
> 
> **First client**
> 
> 1. First client has rw access to __manage namespace. it can see all enrollments and __manage keys
> 
> ```
> @alice@enroll:list
> data:{"1a901b8e-0e89-4b09-9e94-d7b06a12d818.new.enrollments.__manage@alice":{"appName":"buzz","deviceName":"pixel","namespace":{"buzz":"rw"}},"e8e69f4a-2336-4f90-8235-f0a66b5713db.new.enrollments.__manage@alice":{"appName":"wavi","deviceName":"pixel","namespace":{"wavi":"rw","__manage":"rw"}}}
> @alice@keys:get:self
> data:["1a901b8e-0e89-4b09-9e94-d7b06a12d818.default_self_enc_key.__manage@alice","e8e69f4a-2336-4f90-8235-f0a66b5713db.default_self_enc_key.__manage@alice"]
> @alice@keys:get:private
> data:["1a901b8e-0e89-4b09-9e94-d7b06a12d818.default_enc_private_key.__manage@alice","e8e69f4a-2336-4f90-8235-f0a66b5713db.default_enc_private_key.__manage@alice"]
> ```

@gkc please review these steps and whether the output is as expected per the spec

**Comment by gkc** [comment]
This all looks correct to me, @murali-shris

**Comment by murali-shris** [comment]
Pending server tasks/features not in the initial PR 
https://github.com/atsign-foundation/at_server/pull/1495/files
* OTP improvements 
* Expiry of enroll requests which have not been approved/denied for a configured time interval
* Rate limiting of enroll requests - have to create a new task
* override enrollment, change namespace access - have to create a new task

**Comment by murali-shris** [comment]
Impl for spec changes, unit/functional tests done. PR under review
https://github.com/atsign-foundation/at_server/pull/1496

**Comment by sitaram-kalluri** [comment]
Work done in PR-68

Changes to LookUp verb and updating the unit/functional tests as suggested by Gary in last sprint. Spent 3 SP on updating the tests and 5 SP on the lookup_verb changes

**Comment by sitaram-kalluri** [comment]
The changes are merged to trunk branch.

**Comment by sitaram-kalluri** [comment]
The changes are released into Canary v3.0.35a. Closing this issue.
